pkg/storegateway/series_chunks.go

```diff
@@ -417,6 +422,183 @@ func (c *loadingSeriesChunksSetIterator) Err() error {
+			dst[cIdx].Raw = &storepb.Chunk{}
+		}
+	}
+	ok, lastChunkLen, totalRead, err := parseRange(rawRange, dst)
+	if err != nil {
+		return false, err
+	}
+	if !ok {
+		// We estimate the length of the last chunk of a series.
+		// Unfortunately, we got it wrong. We can set the length correctly because we now know it.
+		r.refs[len(r.refs)-1].length = lastChunkLen
+		return false, nil
+	}
+	s.rawRanges[rIdx] = s.rawRanges[rIdx][:totalRead]
+	return true, nil
+}
+
+// parseRange parses the byte slice as concatenated encoded chunks. lastChunkLen is non-zero when allChunksComplete==false.
+// An error is returned when gBytes are malformed or when more than the last chunk is incomplete.
+//
+//nolint:unused // dead code while we are working on PR 3968
+func parseRange(rBytes []byte, chunks []storepb.AggrChunk) (allChunksComplete bool, lastChunkLen uint32, totalRead int, _ error) {
+	rangeFullSize := len(rBytes)
+	for i := range chunks {
+		// ┌───────────────┬───────────────────┬──────────────┬────────────────┐
+		// │ len <uvarint> │ encoding <1 byte> │ data <bytes> │ CRC32 <4 byte> │
+		// └───────────────┴───────────────────┴──────────────┴────────────────┘
+		chunkDataLen, n := varint.Uvarint(rBytes)
+		if n == 0 {
+			return false, 0, 0, fmt.Errorf("not enough bytes (%d) to read length of chunk %d/%d", len(rBytes), i, len(chunks))
+		}
+		if n < 0 {
+			return false, 0, 0, fmt.Errorf("chunk length doesn't fit into uint64 %d/%d", i, len(chunks))
+		}
+		totalChunkLen := n + 1 + int(chunkDataLen) + crc32.Size
+		if totalChunkLen > len(rBytes) {
+			if i != len(chunks)-1 {
+				return false, 0, 0, fmt.Errorf("underfetched before the last chunk, don't know what to do (chunk idx %d/%d, fetched %d/%d bytes)", i, len(chunks), len(rBytes), totalChunkLen)
+			}
+			return false, uint32(totalChunkLen), rangeFullSize - len(rBytes), nil
+		}
+		c := rawChunk(rBytes[n : n+1+int(chunkDataLen)])
+		enc, ok := convertChunkEncoding(c.Encoding())
+		if !ok {
+			return false, 0, 0, fmt.Errorf("unknown encoding (%d, %s), don't know what to do", c.Encoding(), c.Encoding().String())
+		}
+		chunks[i].Raw.Type = enc
+		chunks[i].Raw.Data = c.Bytes()
+		// We ignore the crc32 because we assume that the chunk didn't get corrupted in transit or at rest.
+		rBytes = rBytes[totalChunkLen:]
+	}
+	return true, 0, rangeFullSize - len(rBytes), nil
+}
+
+// parseRanges parses the passed bytes into nextSet. In case a range was underfetched, parseRanges will return an underfetchedChunksRangeIdx
+// with the indices of the range; the keys in the map are the indices into partialSeries.
+// parseRanges will also set the correct length of the last chunk in an underfetched range in case its estimated length was wrong.
+//
+//nolint:unused // dead code while we are working on PR 3968
+func parseRanges(partialSeries []partialSeriesChunksSet) (map[int][]underfetchedChunksRangeIdx, error) {
+	underfetchedSeries := make(map[int][]underfetchedChunksRangeIdx)
```

Comment:
Have you considered lazy creating this map and returning `nil` if there are no underfetched chunk ranges? I think having underfetched chunks is very uncommon.